File: src/backend/lyrics/LrcLib.ts
import { GenericLyrics, GenericLyricsLine } from "../../common/types/LyricsTypes"
import { GenericSong } from "../../common/types/SongTypes"

const LRCLIB_LYRICS_BASE_URL = 'https://lrclib.net/api/get'
const LRCLIB_USER_AGENT = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/141.0.0.0 Safari/537.36 wisp/1.0.0'

type LrcLibLyrics = {
    id: string,
    trackName: string,
    artistName: string,
    albumName: string,
    duration: number,
    instrumental: boolean,
    plainLyrics: string,
    syncedLyrics: string
}

class LrcLibProvider {
    async getLyrics(song: GenericSong) {
        const artistName = encodeURIComponent(song.artists.map(artist => artist.name).join(", "))
        const trackName = encodeURIComponent(song.title)
        const albumName = encodeURIComponent(song.album ? encodeURIComponent(song.album.title) : "")
        const songDuration = song.durationSecs

        const URL = `${LRCLIB_LYRICS_BASE_URL}?artist_name=${artistName}&track_name=${trackName}&album_name=${albumName}&duration=${songDuration}`
        
        const response = await fetch(URL, {
            headers: {
                'Accept': 'application/json',
                'User-Agent': LRCLIB_USER_AGENT,
            },
        })

        try {
            if (response.ok) {
                const lyricsResponse = (await response.json()) as LrcLibLyrics
                const synced = (lyricsResponse.syncedLyrics && lyricsResponse.syncedLyrics.length > 0) ? true : false

                switch (synced) {
                    case true: {
                        // Parse synced lyrics format: [mm:ss.cc] Lorem ipsum dolor sit amet
                        const lines: GenericLyricsLine[] = lyricsResponse.syncedLyrics.split("\n")
                            .filter((line: string) => line.trim())
                            .map((line: string) => {
                                // Match timestamp pattern [mm:ss.cc]
                                const match = line.match(/\[(\d{2}):(\d{2})\.(\d{2})\]\s*(.*)/)
                                if (match) {
                                    const minutes = parseInt(match[1])
                                    const seconds = parseInt(match[2])
                                    const centiseconds = parseInt(match[3])
                                    const content = match[4]
                                    
                                    // Convert to milliseconds
                                    const startTimeMs = (minutes * 60 * 1000) + (seconds * 1000) + (centiseconds * 10)
                                    
                                    return {
                                        content: content,
                                        startTimeMs: startTimeMs.toString()
                                    }
                                }
                                // Fallback if pattern doesn't match
                                return {
                                    content: line,
                                    startTimeMs: "0"
                                }
                            })
                        
                        const properLyrics: GenericLyrics = {
                            provider: "lrclib",
                            synced: synced,
                            lines: lines
                        }

                        return properLyrics
                    }
                    case false: {
                        // Parse plain lyrics - just split by newline
                        const lines: GenericLyricsLine[] = lyricsResponse.plainLyrics.split("\n")
                            .filter((line: string) => line.trim())
                            .map((line: string) => ({
                                content: line,
                                startTimeMs: "0"
                            }))
                        
                        const properLyrics: GenericLyrics = {
                            provider: "lrclib",
                            synced: synced,
                            lines: lines
                        }

                        return properLyrics
                    }
                }
            }
        } catch (e) {
            console.log(e)
        }
    }
}

export const lrcLibProvider = new LrcLibProvider()

File: lib/providers/lyrics/lrclib.dart
/// LrcLib lyrics provider
library;

import 'dart:convert';
import 'package:http/http.dart' as http;
import '../../models/metadata_models.dart';

const _lrclibBaseUrl = 'https://lrclib.net/api/get';
const _lrclibUserAgent =
		'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
		'(KHTML, like Gecko) Chrome/141.0.0.0 Safari/537.36 wisp/1.0.0';

class LrcLibLyricsProvider {
	Future<LyricsResult?> getLyrics(GenericSong song, LyricsSyncMode mode) async {
		final artistName = song.artists.map((artist) => artist.name).join(', ');
		final trackName = song.title;
		final albumName = song.album?.title ?? '';

		final uri = Uri.parse(_lrclibBaseUrl).replace(
			queryParameters: {
				'artist_name': artistName,
				'track_name': trackName,
				'album_name': albumName,
				'duration': song.durationSecs.toString(),
			},
		);

		final response = await http.get(
			uri,
			headers: {
				'Accept': 'application/json',
				'User-Agent': _lrclibUserAgent,
			},
		);

		if (response.statusCode != 200) return null;

		try {
			final data = jsonDecode(response.body) as Map<String, dynamic>;
			final syncedLyrics = (data['syncedLyrics'] as String?) ?? '';
			final plainLyrics = (data['plainLyrics'] as String?) ?? '';

			final hasSynced = syncedLyrics.trim().isNotEmpty;
			final hasPlain = plainLyrics.trim().isNotEmpty;

			if (mode == LyricsSyncMode.synced && hasSynced) {
				final lines = _parseSyncedLyrics(syncedLyrics);
				return LyricsResult(
					provider: LyricsProviderType.lrclib,
					synced: true,
					lines: lines,
				);
			}

			if (mode == LyricsSyncMode.unsynced && hasPlain) {
				final lines = _parsePlainLyrics(plainLyrics);
				return LyricsResult(
					provider: LyricsProviderType.lrclib,
					synced: false,
					lines: lines,
				);
			}

			if (hasSynced) {
				final lines = _parseSyncedLyrics(syncedLyrics);
				return LyricsResult(
					provider: LyricsProviderType.lrclib,
					synced: true,
					lines: lines,
				);
			}

			if (hasPlain) {
				final lines = _parsePlainLyrics(plainLyrics);
				return LyricsResult(
					provider: LyricsProviderType.lrclib,
					synced: false,
					lines: lines,
				);
			}
		} catch (_) {
			return null;
		}

		return null;
	}

	List<LyricsLine> _parseSyncedLyrics(String syncedLyrics) {
		return syncedLyrics
				.split('\n')
				.where((line) => line.trim().isNotEmpty)
				.map((line) {
			final match = RegExp(r'\[(\d{2}):(\d{2})\.(\d{2})\]\s*(.*)')
					.firstMatch(line);
			if (match != null) {
				final minutes = int.parse(match.group(1)!);
				final seconds = int.parse(match.group(2)!);
				final centiseconds = int.parse(match.group(3)!);
				final content = match.group(4) ?? '';
				final startTimeMs =
						(minutes * 60 * 1000) + (seconds * 1000) + (centiseconds * 10);
				return LyricsLine(content: content, startTimeMs: startTimeMs);
			}
			return LyricsLine(content: line, startTimeMs: 0);
		}).toList();
	}

	List<LyricsLine> _parsePlainLyrics(String plainLyrics) {
		return plainLyrics
				.split('\n')
				.where((line) => line.trim().isNotEmpty)
				.map((line) => LyricsLine(content: line, startTimeMs: 0))
				.toList();
	}
}
